src/sensors/mod.rs

```diff
@@ -0,0 +1,93 @@
+//! Module defining common sensor functionality.
+
+mod humidity;
+mod people_now_present;
+mod power_consumption;
+mod temperature;
+
+pub use humidity::{HumiditySensor, HumiditySensorTemplate};
+pub use people_now_present::{PeopleNowPresentSensor, PeopleNowPresentSensorTemplate};
+pub use power_consumption::{PowerConsumptionSensor, PowerConsumptionSensorTemplate};
+pub use temperature::{TemperatureSensor, TemperatureSensorTemplate};
+
+use log::warn;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+/// Common information describing any sensor.
+#[derive(Serialize, Deserialize, Default, Debug, Clone, PartialEq)]
+pub struct SensorMetadata {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub name: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub location: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub description: Option<String>,
+}
+
+/// Common information describing any sensor which requires a specified location.
+#[derive(Serialize, Deserialize, Default, Debug, Clone, PartialEq)]
+pub struct LocalisedSensorMetadata {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub name: Option<String>,
+    pub location: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub description: Option<String>,
+}
+
+/// Describes an error occurring when building a sensor from a `SensorTemplate`.
+#[derive(Error, Debug)]
+pub enum SensorTemplateError {
+    /// Failed when parsing an integer value from the provided value string
+    #[error("sensor integer value cannot be parsed")]
+    BadInteger(#[from] std::num::ParseIntError),
+
+    /// Failed when parsing a floating point value from the provided value string
+    #[error("sensor float value cannot be parsed")]
+    BadFloat(#[from] std::num::ParseFloatError),
+}
+
+/// A trait for all possible sensor templates.
+///
+/// A sensor template is like a sensor struct, but without the actual data in it.
+/// A `SensorTemplate` is capable of registering itself in a `Sensors` struct.
+pub trait SensorTemplate: Send + Sync {
+    fn to_sensor(&self, value_str: &str, sensors: &mut Sensors) {
```

Comment:
General observation: The `&str` is a bit ugly and should probably be replaced by generics or a custom `Value` enum in the long term. However, it's fine for now. (I created this API when I knew much less about Rust.)

Reply:
I think the `&str` comes from us using redis in `spaceapi-server.rs` to store the values which only uses strings.

Maybe we could use https://doc.rust-lang.org/std/convert/trait.TryInto.html? But this would make the error type generic.

Reply:
redis-rs supports other datatypes as well. But we probably used strings because we don't know the type of the field in advance without a schema, so a string probably makes sense.

Nevertheless, "spaceapi-rs" could use proper types, and the server could convert from strings to those types in advance 🙂 But that's a thing for a future PR, so I'll resolve this comment.